### PR TITLE
[FLINK-1979] Lossfunctions

### PR DESCRIPTION
**Pull request by jojo19893**
We added Logistic Loss Functions and Hinge Loss to the Optimazation Framework.
See for the implemented Functions: 
https://github.com/JohnLangford/vowpal_wabbit/wiki/Loss-functions

Jira Issue: 
https://issues.apache.org/jira/browse/FLINK-1979

### CODE REVIEW

**Comment by thvasilo** [comment]
Thank you for the contribution Johaness. For the logistic loss could you change the implementation to match the approximate version that sklearn (and other libraries) use?

You can see it [here](https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/linear_model/sgd_fast.pyx#L202), this way we may save the computation of the log.
@tillrohrmann Any objections to using this version?

**Comment by jojo19893** [comment]
Okay I Changed the implementation of the logistic loss! Thanks for the advice !

**Comment by thvasilo** [comment]
Thank you Johaness. The optimization code has been merged to the master now, so could you rebase your branch to the latest master so we can look at the changes in an isolated way?

You can take a look at the [How to contibute](http://flink.apache.org/how-to-contribute.html#contributing-code-&-documentation) guide on how to do this. The merges you have currently make it hard to review the code.

Also, please make sure all your classes have docstrings, you can take the docstring for SquaredLoss as an example (i.e. one sentence is usually enough).

Documentation is always welcome of course, so if you want to add some more details to the loss functions section of the ML documentation (docs/libs/ml/optimization.md) feel free to do so in this PR.

Let me know if you run into any problems.

**Comment by aalexandrov** [comment]
This needs a PR before it is merged. @jojo19893, @mguldner please cherry pick your changes such that you have only one or two commits and force push in this branch as suggested by @tillrohrmann in the JIRA.

**Comment by thvasilo** [comment]
Hello, this PR needs to be rebase since so that it's mergeable with the latest master.

I think it's fair to say that we can close it in a week if no more activity occurs.

**Comment by skavulya** [comment]
I need this pull request for work I am doing with L-BFGS. I would like to take it over and make the recommended changes since there has not been any activity on this for a while.

**Comment by tillrohrmann** [comment]
Sure go ahead @skavulya.

**Comment by skavulya** [comment]
@tillrohrmann I updated the loss functions and regularization penalties based on your branch. I was not sure whether to update the gradient descent algorithm to pass the regularization penalty as a parameter in this branch so I kept the GradientDescent API as-is. 

Let me know if you would like me to change the GradientDescent API to GradientDescent().setRegularizationPenalty(L1Regularization) and I will update my branch, squash the commits, and create a PR. https://github.com/apache/flink/compare/master...skavulya:loss-functions

**Comment by thvasilo** [comment]
Hello @skavulya thank you for your contribution, I'm looking forward to the PR!

Re. the regularization penalty, we it changed to make user choice easier, e.g. preventing the use of an L1 penalty when L-BFGS is used.

Personally I would prefer to go back to the original design and have it as a settable parameter, and apply some sanity checking within the optimizer to avoid cases like the one mentioned above.

I would suggest you open the PR and we can discuss it there.

**Comment by skavulya** [comment]
thanks @thvasilo! I created the PR https://github.com/apache/flink/pull/1985

**Comment by zentol** [comment]
Subsumed in #1985.
